go/vt/vttablet/onlineddl/analysis.go

```diff
@@ -0,0 +1,209 @@
+/*
+Copyright 2022 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package onlineddl
+
+import (
+	"context"
+	"encoding/json"
+
+	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/schema"
+	"vitess.io/vitess/go/vt/sqlparser"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+type specialAlterOperation string
+
+const (
+	dropRangePartitionSpecialOperation specialAlterOperation = "drop-range-partition"
+	addRangePartitionSpecialOperation  specialAlterOperation = "add-range-partition"
+)
+
+type SpecialAlterPlan struct {
+	operation   specialAlterOperation
+	details     map[string]string
+	alterTable  *sqlparser.AlterTable
+	createTable *sqlparser.CreateTable
+}
+
+func NewSpecialAlterOperation(operation specialAlterOperation, alterTable *sqlparser.AlterTable, createTable *sqlparser.CreateTable) *SpecialAlterPlan {
+	return &SpecialAlterPlan{
+		operation:   operation,
+		details:     map[string]string{"operation": string(operation)},
+		alterTable:  alterTable,
+		createTable: createTable,
+	}
+}
+
+func (p *SpecialAlterPlan) SetDetail(key string, val string) *SpecialAlterPlan {
+	p.details[key] = val
+	return p
+}
+func (p *SpecialAlterPlan) Detail(key string) string {
+	return p.details[key]
+}
+
+func (p *SpecialAlterPlan) String() string {
+	b, err := json.Marshal(p.details)
+	if err != nil {
+		return ""
+	}
+	return string(b)
+}
+
+// getCreateTableStatement gets a formal AlterTable representation of the given table
+func (e *Executor) getCreateTableStatement(ctx context.Context, tableName string) (*sqlparser.CreateTable, error) {
+	showCreateTable, err := e.showCreateTable(ctx, tableName)
+	if err != nil {
+		return nil, err
+	}
+	stmt, err := sqlparser.ParseStrictDDL(showCreateTable)
+	if err != nil {
+		return nil, err
+	}
+	createTable, ok := stmt.(*sqlparser.CreateTable)
+	if !ok {
+		return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "expected CREATE TABLE. Got %v", sqlparser.CanonicalString(stmt))
+	}
+	return createTable, nil
+}
+
+// analyzeDropFirstOrLastRangePartition sees if the online DDL drops the first or last partition
+//  in a range partitioned table
```

Comment:
Isn't this only checking if it's the first partition? 

And isn't adding checking if it's adding the last? So is this comment here correct?

Reply:
fixed. The code was refactored into a more general form, and the original comment was left behind.